gift-certificate-application/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java

```diff
@@ -0,0 +1,301 @@
+        giftCertificateList = new ArrayList<>();
+        giftCertificateList.add(giftCertificate);
+        tagNames = tagList.stream().map(tagName -> tag.getName()).collect(Collectors.toList());
+        columnNames = new ArrayList<>(Collections.singleton(ANY_STRING));
+    }
+
+    @Test
+    void testCreate() {
+        GiftCertificate expectedResult = giftCertificate;
+        when(tagDao.findByName(TAG_NAME)).thenReturn(Optional.ofNullable(tag));
+        when(tagDao.create(tag)).thenReturn(tag);
+        when(tagValidator.isNameValid(TAG_NAME)).thenReturn(true);
+        when(giftCertificateValidator.validateAll(NAME, DESCRIPTION, PRICE, DURATION)).thenReturn(true);
+        when(giftCertificateDao.findByName(NAME)).thenReturn(Optional.empty());
+        when(giftCertificateDao.create(giftCertificate)).thenReturn(giftCertificate);
+        GiftCertificate actualResult = giftCertificateService.create(giftCertificate);
+        assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    void testCreateShouldThrowFieldValidationExceptionWhenTagInvalid() {
+        when(tagValidator.isNameValid(anyString())).thenReturn(false);
+        when(giftCertificateValidator.validateAll(anyString(), anyString(), any(), anyShort())).thenReturn(true);
+        assertThrows(FieldValidationException.class, () -> giftCertificateService.create(giftCertificate));
+    }
+
+    @Test
+    void testCreateShouldThrowFieldValidationExceptionWhenGiftCertificateInvalid() {
+        when(tagValidator.isNameValid(anyString())).thenReturn(true);
+        when(giftCertificateValidator.validateAll(anyString(), anyString(), any(), anyShort())).thenReturn(false);
+        assertThrows(FieldValidationException.class, () -> giftCertificateService.create(giftCertificate));
+    }
+
+    @Test
+    void testCreateShouldThrowGiftCertificateAlreadyExistException() {
+        when(tagValidator.isNameValid(anyString())).thenReturn(true);
+        when(giftCertificateValidator.validateAll(anyString(), anyString(), any(), anyShort())).thenReturn(true);
+        when(giftCertificateDao.findByName(anyString())).thenReturn(Optional.of(giftCertificate));
+        assertThrows(GiftCertificateAlreadyExistException.class,
+                () -> giftCertificateService.create(giftCertificate));
+    }
+
+    @Test
+    void testFindAll() {
+        List<GiftCertificate> expectedResult = giftCertificateList;
+        when(giftCertificateDao.findAll(pagination, GiftCertificate.class)).thenReturn(giftCertificateList);
+        when(giftCertificateDao.findEntitiesNumber(GiftCertificate.class)).thenReturn(ENTITIES_NUMBER);
+        when(paginationValidator.validatePagination(pagination, ENTITIES_NUMBER)).thenReturn(pagination);
+        List<GiftCertificate> actualResult = giftCertificateService.findAll(pagination);
+        assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    void testFindById() {
+        GiftCertificate expectedResult = giftCertificate;
```

Comment:
I suppose you can directly use giftCertificate and do not create an additional variable